---
title: "4. Available widgets"
author:
- name: Martin Morgan
  affiliation: Roswell Park Cancer Insitute, Buffalo, NY, USA
  email: Martin.Morgan@RoswellPark.org
- name: Marcel Ramos
  affiliation: CUNY School of Public Health, New York, NY, USA
package: SOUNDBoard
output:
  SOUNDBoard::html_document:
    toc: true
    toc_depth: 2
runtime: shiny
vignette: >
  %\VignetteIndexEntry{3. Creating and deploying a report}
  %\VignetteEngine{knitr::rmarkdown}
---

Types of widgets

1. Static -- e.g., presenting static information, e.g., patient id,
   case number, ...

2. Interactive widgets -- JavaScript-based interactions for displaying
   static information that the user can manipulate, e.g., large
   tables, interactive plots.

3. shiny-based widgets -- Require evaluation in R, doing arbitrary
   'read-only' operations that could potentially be time-consuming.

4. Linked widgets -- e.g., [crosstalk][] -- choice of gene in table of
   variants updates choice of available drugs.

# Generic implementations

Status: planned, beta, available

## Static

- Serialized _R_ objects, other than `data.frame()` or `ggplot()`. _beta_

    ```
    ## widget <- ... # any R object
    tbl(x, "assay") <- list(
        case_uid = "PATIENT_1",
        assay = "R object",
        description = "Demonstrating R object management",
        resource = manage(x, widget)
    )
    ```

- `tbl_board` and `tbl_cases`. _beta_

    ```
    tbl(x, "board") <- list(
        board_uid = "Unique_Board_ID"
        description = "Board description"
    )
    ```
    
    ```
    tbl(x, "cases") <- list(
        board_uid = "Unique_Board_ID",
        case_uid = "Unique_Case_ID",
        age = 36,
        sex = "Male"
    )
    ```

## Interactive

- `data.frame()` via `DT::datatable()`. _beta_

    ```
    ## assay resource is a data.frame object
    ## tbl(x, "assay") <- list(..., assay = "data.frameObj",
    ##    resource = manage(x, ggplot2::ggplot(...))) # save into assays list
    tbl(x, "assay") %>%
        filter(case_uid == "Unique_Case_ID", assay == "data.frameObj") %>%
        sbreport()
    ```

- `ggplot()` via `plotly::ggplotly()`. _beta_
    
    ```
    ## assay resource is a ggplot object
    ## tbl(x, "assay") <- list(..., assay = "ggplotObj",
    ##    resource = manage(x, ggplot2::ggplot(...))) # save into assays list
    tbl(x, "assay") %>%
        filter(case_uid == "Unique_Case_ID", assay == "ggplotObj") %>%
        sbreport()
    ```

## Shiny-based

- `ShinyAppWidget()`. _beta_ 
  -specified via _R_ code in a single file.
  
    ```
    ## myApp <- ShinyAppWidget(...) # assign shiny app as object
    ## tbl(x, "assay") <- list(..., assay = "myApp") # save into assays list
    tbl(x, "assay") %>%
        filter(case_uid == "Unique_Case_ID", assay == "myApp") %>%
        sbreport()
    ```

- `ShinyAppDirWidget()`. _beta_
  - _R_ code in ui.R / server.R files.

    ```
    ## myShinyDir <- ShinyAppDirWidget(resource = "path/to/app/dir")
    ## tbl(x, "assay") <- list(..., assay = "shinyAppDir",
    ##    resource = manage(x, myShinyDir)) # save into assays
    tbl(x, "assay") %>%
        filter(case_uid == "Unique_Case_ID", assay == "shinyAppDir")
    ```

## Linked

# Specific implementations

- `HeatmapWidget()`. _beta_
  - rows: genes; columns: samples
  - samples clustered; bars to indicate phenotype
  - interactive in the sense that hover over a cell provides row,
    column, value info
  - implemented with [ComplexHeatmap][]
    
    ```
    ## Make a Heatmap using the ComplexHeatmap package
    ## heatmap <- ComplexHeatmap::Heatmap(matrix(...)) 
    widget <- HeatmapWidget(heatmap)
    tbl(x, "assay") <- list(
        case_uid = "PATIENT_1",
        assay = "R object",
        description = "Demonstrating R object management",
        resource = manage(x, widget)
    )
    ```

- `OncoPrintWidget()`. _beta_
  - rows: genes: columns: samples
  - Visualizes multiple genomic alteration events via heatmap
  - implemented with [ComplexHeatmap][]

    ```
    ## Use the ComplexHeatmap to make an oncoPrint graphic
    ht <- ComplexHeatmap::oncoPrint(...)
    tbl(x, "assay") <- list(
        case_uid = "TCGA-32-2615",
        assay = "onco",
        description = "Heatmap Widget Demo",
        resource = manage(x, ht)
    )
    ```

- `DGIdbWidget()`. _planned_
  - query DGIdb with genes, implemented with
    [rDGIdb][]

[ComplexHeatmap]: https://bioconductor.org/packages/ComplexHeatmap
[rDGIdb]: https://bioconductor.org/packages/rDGIdb
[crosstalk]: https://rstudio.github.io/crosstalk/
